src/test/java/cloud/fogbow/ras/core/plugins/interoperability/cloudstack/publicip/v4_9/CloudStackPublicIpPluginTest.java

```diff
@@ -0,0 +1,798 @@
+
+    // test case: When calling the finishAsyncRequestInstanceSteps method, it must verify if It
+    // sets the right values in the AsyncRequestInstanceState.
+    @Test
+    public void testFinishAsyncRequestInstanceStepsSuccessfully() {
+        // set up
+        AsyncRequestInstanceState asyncRequestInstanceState = new AsyncRequestInstanceState(null, null , null);
+
+        // verify before
+        Assert.assertNull(asyncRequestInstanceState.getState());
+
+        // exercise
+        this.plugin.finishAsyncRequestInstanceSteps(asyncRequestInstanceState);
+
+        // verify after
+        Assert.assertEquals(AsyncRequestInstanceState.StateType.READY, asyncRequestInstanceState.getState());
+    }
+
+    // test case: When calling the setAsyncRequestInstanceSecondStep method, it must verify if It
+    // sets the right values in the AsyncRequestInstanceState.
+    @Test
+    public void testSetAsyncRequestInstanceSecondStepSuccessfully() {
+        // set up
+        String ipAddressIdExpected = "ipId";
+        String ipExpected = "ip";
+
+        SuccessfulAssociateIpAddressResponse response = Mockito.mock(SuccessfulAssociateIpAddressResponse.class);
+        SuccessfulAssociateIpAddressResponse.IpAddress ipAddress =
+                Mockito.mock(SuccessfulAssociateIpAddressResponse.IpAddress.class);
+        Mockito.when(ipAddress.getId()).thenReturn(ipAddressIdExpected);
+        Mockito.when(ipAddress.getIpAddress()).thenReturn(ipExpected);
+
+        Mockito.when(response.getIpAddress()).thenReturn(ipAddress);
+        AsyncRequestInstanceState asyncRequestInstanceState = new AsyncRequestInstanceState(null, null , null);
+        String createFirewallRuleJobId = "jobId";
+
+        // verify before
+        Assert.assertNull(asyncRequestInstanceState.getIpInstanceId());
+        Assert.assertNull(asyncRequestInstanceState.getIp());
+        Assert.assertNull(asyncRequestInstanceState.getState());
+        Assert.assertNull(asyncRequestInstanceState.getCurrentJobId());
+
+        // exercise
+        this.plugin.setAsyncRequestInstanceSecondStep(response, asyncRequestInstanceState, createFirewallRuleJobId);
+
+        // verify after
+        Assert.assertEquals(ipAddressIdExpected, asyncRequestInstanceState.getIpInstanceId());
+        Assert.assertEquals(ipExpected, asyncRequestInstanceState.getIp());
+        Assert.assertEquals(AsyncRequestInstanceState.StateType.CREATING_FIREWALL_RULE,
+                asyncRequestInstanceState.getState());
+        Assert.assertEquals(createFirewallRuleJobId, asyncRequestInstanceState.getCurrentJobId());
+    }
+
+    // test case: When calling the requestCreateFirewallRule method and occurs a HttpResponseException,
+    // it must verify if It throws FogbowException.
+    @Test
+    public void testRequestCreateFirewallRuleFail() throws FogbowException, HttpResponseException {
+        // set up
+        CreateFirewallRuleRequest request = new CreateFirewallRuleRequest.Builder().build("");
```

Comment:
Use the **_EMPTY_STRING_** constant of the **TestUtils** class... Do this for other occurrences.

Reply:
Done.
Note: I've done the changes in the fourth PR in order to avoid conflicts.